PayloadPkg/OsLoader/OsLoader.c

```diff
@@ -1297,6 +1312,41 @@ BootOsImage (
     goto Exit;
   }
 
+  if (MediaType == OsBootDeviceMemory) {
```

Comment:
Instead of add a special handling here, did you try to use Memory Device to load image from memory? Memory device could be found here https://github.com/slimbootloader/slimbootloader/blob/master/BootloaderCommonPkg/Library/MemoryDeviceBlockIoLib/MemoryDeviceBlockIoLib.c.
Memory is similar with memory mapped SPI device. It would be more consistent to use MemoryDeviceBlockIoLib.

Reply:
Just echo Guo: according to my test, the MemoryDeviceBlockIoLib already works well for OsBootDeviceMemory when calling from GetBootImageFromRawPartition().

However, one thing to note is about the MemoryDeviceGetMediaInfo(), it sets DevBlockInfo->BlockNum  = 0x100000000 without checking the real physical memory size. Maybe we should modify MemoryDeviceGetMediaInfo() to reveal the real physical memory size.

Reply:
Updated

Reply:
DevBlockInfo->BlockNum is not used in the boot from memory case.